### PR TITLE
Allow negative fee1 kb low

### PR DESCRIPTION
**Pull request by sisuresh**
### What

Allows for a more aggressive increase in the fee before the target size is hit.

### CODE REVIEW

File: soroban-env-host/src/fees.rs [hunk 1]
```diff
@@ -238,18 +241,23 @@ pub fn compute_write_fee_per_1kb(
...
+            (fee_rate_multiplier as i128)
+                .saturating_mul(bucket_list_size_before_reaching_target as i128),
+            fee_config.bucket_list_target_size_bytes as i128,
+        )
+        .clamp_fee();
+        // no clamp_fee here
+        write_fee_per_1kb =
+            write_fee_per_1kb.saturating_add(fee_config.write_fee_1kb_bucket_list_low);
+    } else {
+        write_fee_per_1kb = fee_config.write_fee_1kb_bucket_list_high;
```
Comment (sisuresh) [inline soroban-env-host/src/fees.rs:260]: From offline discussions - 
Siddharth - "So if we're over the target size, we charge a fixed fee (write_fee_1kb_bucket_list_high) for the bytes up to bucket_list_target_size_bytes? The calculations below only deal with the bytes after the target size. Was this intentional?".

Comment (sisuresh) [reply soroban-env-host/src/fees.rs:260]: Nico - "The CAP has this written in a fairly convoluted way, I agree.
before this change:
```
write_fee_per_1kb(s) = (writeFee1KBBucketListHigh - writeFee1KBBucketListLow)*min(s, bucketListTargetSizeBytes)/bucketListTargetSizeBytes +
  writeFee1KBBucketListLow +
  (if s > bucketListTargetSizeBytes,
      bucketListWriteFeeGrowthFactor*
      (writeFee1KBBucketListHigh - writeFee1KBBucketListLow)*
        (s-bucketListTargetSizeBytes)/bucketListTargetSizeBytes)
```

In plain English, we just have 2 curves (well 3):
* before target, it's a line that starts at `(0, write_fee_1kb_bucket_list_low)` and ends at `(bucket_list_target_size_bytes, write_fee_1kb_bucket_list_high)`
     * as we allow the line to intercept with a low (negative) `write_fee_1kb_bucket_list_low` we clamp it with `MINIMUM_WRITE_FEE_PER_1KB`  -> this creates a first "flat" portion to the first part of the fee curve.
* after target, it's the line that starts at `(bucket_list_target_size_bytes, write_fee_1kb_bucket_list_high)` with a slope of `bucketListWriteFeeGrowthFactor*
      (writeFee1KBBucketListHigh - writeFee1KBBucketListLow)/bucketListTargetSizeBytes`  --> `bucketListWriteFeeGrowthFactor` faster than the slope in step 1"

Comment (sisuresh) [reply soroban-env-host/src/fees.rs:260]: @MonsieurNicolas am I right in saying that the main motivation for this change was to allow for the increase of the slope before the target size is hit without increasing writeFee1KBBucketListHigh? I'm trying to figure out why we can't just increase writeFee1KBBucketListHigh instead. The other important part was setting the min.

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
can we add a few tests for the boundary conditions when there is clamping?

File: soroban-env-host/src/fees.rs [hunk 2]
```diff
@@ -261,7 +266,11 @@ pub fn compute_write_fee_per_1kb(
         .clamp_fee();
         write_fee_per_1kb = write_fee_per_1kb.saturating_add(post_target_fee);
     }
-    write_fee_per_1kb
+    if write_fee_per_1kb < MINIMUM_WRITE_FEE_PER_1KB {
+        MINIMUM_WRITE_FEE_PER_1KB
+    } else {
+        write_fee_per_1kb
+    }
```
Comment (jayz22) [inline soroban-env-host/src/fees.rs:273]: nit: you can just do `write_fee_per_1kb.max(MINIMUM_WRITE_FEE_PER_1KB)`

Comment (sisuresh) [reply soroban-env-host/src/fees.rs]: Good point. Fixed

File: soroban-env-host/src/fees.rs [hunk 3]
```diff
@@ -238,18 +241,20 @@ pub fn compute_write_fee_per_1kb(
...
+            fee_config.bucket_list_target_size_bytes as i128,
+        )
+        .clamp_fee();
+        // no clamp_fee here
+        write_fee_per_1kb =
+            write_fee_per_1kb.saturating_add(fee_config.write_fee_1kb_bucket_list_low);
+    } else {
+        write_fee_per_1kb = fee_config.write_fee_1kb_bucket_list_high;
         let bucket_list_size_after_reaching_target =
             bucket_list_size_bytes - fee_config.bucket_list_target_size_bytes;
```
Comment (graydon) [inline soroban-env-host/src/fees.rs:259]: A little concerned that this subtraction can underflow

File: soroban-env-host/src/fees.rs [hunk 4]
```diff
@@ -238,18 +241,20 @@ pub fn compute_write_fee_per_1kb(
...
-        fee_config.bucket_list_target_size_bytes as i128,
-    )
-    .clamp_fee();
-    write_fee_per_1kb = write_fee_per_1kb.saturating_add(fee_config.write_fee_1kb_bucket_list_low);
-    if bucket_list_size_bytes > fee_config.bucket_list_target_size_bytes {
+    let mut write_fee_per_1kb: i64;
+    if bucket_list_size_bytes < fee_config.bucket_list_target_size_bytes {
+        // Convert multipliers to i128 to make sure we can handle large bucket list
+        // sizes.
+        write_fee_per_1kb = num_integer::div_ceil(
```
Comment (graydon) [inline soroban-env-host/src/fees.rs:248]: This can panic if divisor is zero

File: soroban-env-host/src/fees.rs (hunk 3 above)
Comment (graydon) [inline soroban-env-host/src/fees.rs:259]: this subtraction can underflow

File: soroban-env-host/src/fees.rs [hunk 5]
```diff
@@ -238,18 +241,20 @@ pub fn compute_write_fee_per_1kb(
...
+        )
+        .clamp_fee();
+        // no clamp_fee here
+        write_fee_per_1kb =
+            write_fee_per_1kb.saturating_add(fee_config.write_fee_1kb_bucket_list_low);
+    } else {
+        write_fee_per_1kb = fee_config.write_fee_1kb_bucket_list_high;
         let bucket_list_size_after_reaching_target =
             bucket_list_size_bytes - fee_config.bucket_list_target_size_bytes;
         let post_target_fee = num_integer::div_ceil(
```
Comment (graydon) [inline soroban-env-host/src/fees.rs:260]: this division can panic with zero divisor

**Review by graydon — CHANGES_REQUESTED** [review]
I don't see anything specifically wrong with this _change_ but the code it's changing in general has a subtraction that can underflow and (scanning the file) I think 4 divs that can panic with zero divisors. Can we fix those?

File: soroban-env-host/src/fees.rs (hunk 3 above)
Comment (sisuresh) [reply soroban-env-host/src/fees.rs:259]: This subtraction can't underflow because it's in the else condition of `if bucket_list_size_bytes < fee_config.bucket_list_target_size_bytes`. I can still add the saturating_sub call to future proof if you think that makes sense.

File: soroban-env-host/src/fees.rs (hunk 4 above)
Comment (sisuresh) [reply soroban-env-host/src/fees.rs:248]: I added a .max(1) to all denominators.

File: soroban-env-host/src/fees.rs (hunk 3 above)
Comment (graydon) [reply soroban-env-host/src/fees.rs:259]: yeah, for future proofing I'd appreciate it. thanks!

Comment (sisuresh) [reply soroban-env-host/src/fees.rs]: Done!

**Review by graydon — APPROVED** [review]

**Review by MonsieurNicolas — APPROVED** [review]

File: soroban-env-host/src/fees.rs [hunk 6]
```diff
@@ -13,6 +13,9 @@ pub const TTL_ENTRY_SIZE: u32 = 48;
 const INSTRUCTIONS_INCREMENT: i64 = 10000;
 const DATA_SIZE_1KB_INCREMENT: i64 = 1024;
 
+// minimum effective write fee per 1KB
+pub const MINIMUM_WRITE_FEE_PER_1KB: i64 = 1000;
```
Comment (dmkozh) [inline soroban-env-host/src/fees.rs:17]: I wonder, why not 1 or even 0? This adds some non-trivial logic to already complex fee logic... I guess it's too late to change this, just curious as to whether this was intentional.

Comment (MonsieurNicolas) [reply soroban-env-host/src/fees.rs:17]: yes, this guardrail matches `InitialSorobanNetworkConfig::BUCKET_LIST_FEE_1KB_BUCKET_LIST_LOW` Soroban default that is supposed to be the "lowest sane value" for pubnet.
We can always rewrite the formula later as this should really never trigger in pubnet.
